op/reboot_decide.go

```diff
@@ -0,0 +1,265 @@
+package op
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cybozu-go/cke"
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/well"
+	corev1 "k8s.io/api/core/v1"
+	policyv1 "k8s.io/api/policy/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+)
+
+// enumeratePods enumerates Pods on a specified node.
+// It calls podHandler for each Pods not owned by Job nor DaemonSet and calls jobPodHandler for each running Pods owned by a Job.
+// If those handlers returns error, this function returns the error immediately.
+func enumeratePods(ctx context.Context, cs *kubernetes.Clientset, node string,
+	podHandler func(pod *corev1.Pod) error, jobPodHandler func(pod *corev1.Pod) error) error {
+
+	podList, err := cs.CoreV1().Pods(corev1.NamespaceAll).List(ctx, metav1.ListOptions{
+		FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": node}).String(),
+	})
+	if err != nil {
+		return err
```

Comment:
The error handling of `enumeratePods()` seems somewhat dangerous because the function mixes the real errors and the statuses of unsatisfied conditions.
The callers of this function cannot distinguish the real errors from condition-unsatisfied situations.